Changes for arduinostm32 4.108

sjasonsmith opened the pull request:
### Description

Update to the latest version of the arduinostm32 framework, used by HAL/STM32.

**I have only tested this with an SKR Pro, and only on my bench. I don't currently have it in a printer. I have verified that Servo output, flash access, step generation, temperature reading, and SoftwareSerial function. I have not verified every possible function of the hardware.**

Unfortunately there are some timer-related issues in the current framework, which prevents us from deleting our SoftwareSerial implementation. I have added comments to the workarounds which will hopefully be able to be removed after the next release.

The two issues in the framework are:
1. Must call setMode on a timer channel, even if you are using the timer for simple overflows. This prevents the framework's SoftwareSerial from working at all. This has already been fixed, but not yet released:
https://github.com/stm32duino/Arduino_Core_STM32/pull/849

2. Setting interrupt priority through SoftwareSerial or HardwareTimer entrypoints doesn't apply the priority to already initialized timers. I have submitted a PR to fix this, but I have a little more work to do and we will then need a new release of the framework.
https://github.com/stm32duino/Arduino_Core_STM32/pull/1062

### Benefits

Two key benefits:
1. All HAL/STM32 environments use the same framework version.
2. We can begin identifying additional issues which may exist as people begin using this.

### Related Issues

N/A

sjasonsmith committed:
Update to arduinostm32 4.10800. Update timer usage accordingly.

sjasonsmith commented:
This impacts every board using HAL/STM32. I'm hoping people with something _other_ than an SKR Pro will be able to try it out.

I'm actually most interested in whether it is working with Malyan boards, so maybe @xC0000005 or @mojocorp would be interested in trying it out.

With regard to the Malyan M300, it was already using this framework version because that is where the board variant exists. I have to assume that means its timers were not working prior to these changes, and hopefully work now. I might be way off on those assumptions, since I don't have that hardware to test anything with.

sjasonsmith commented on Marlin/src/HAL/STM32/SoftwareSerial.cpp:
```diff
@@ -137,12 +142,15 @@ void SoftwareSerial::setSpeed(uint32_t speed) {
       timer.setCount(0);
       timer.attachInterrupt(&handleInterrupt);
       timer.resume();
-      NVIC_SetPriority(timer_interrupt_number, timer_interrupt_priority);
```
The required framework required us to reset the priority every time we resumed the timer, but that is no longer necessary. Setting it once in `setInterruptPriority` is adequate.

thinkyhead committed:
arduino_ver

thinkyhead commented:
It's good to see that the timers are getting sorted out. You're correct about M300. Compiles fine but then 👎 .

GhostlyCrowd commented:
I have a SKR Pro 1.1 in a machine, i just pulled this I'm willing to run a few prints make sure everything is as expected. 

Let me know if their is anything specific you want me to do.

sjasonsmith commented:
> I have a SKR Pro 1.1 in a machine, i just pulled this I'm willing to run a few prints make sure everything is as expected.
> 
> Let me know if their is anything specific you want me to do.

Just normal use would be great. Heaters, fans, and any other peripherals you normally use.

sjasonsmith commented:
@thinkyhead I probably should have marked this as a draft, but I don't think I I can change that after-the-fact.

@GhostlyCrowd said in Discord that he might be seeing some issues. We will definitely want testing on this by a few people before it goes in.

GhostlyCrowd commented:
@sjasonsmith  
Just to collect my findings in the correct area for every one to see

Ok so the issues ive found is the stepper movement has a slight very slight almost not noticable but you can feel it and see it in the first layer stutter. Its causing micro extrusion issues as well as the first layer you can see slight defect in the outlines. If I place my hands on my 2020 rails on the side of the Y axis you can feel it, it's just an ever so slight sputtering. 

Second is the speaker will sometimes function correctly (plays a failure musical tone when I cancel the print) and other times it will just chirp once instead of playing the whole musical tone. Maybe timer silliness?

SKR Pro, TMX2209's,  Shut scurve off just top be sure since i know its experimental on Lin adv, still the micro stuttering.  

Also reverted a recent merge for ```#define MINIMUM_STEPPER_POST_DIR_DELAY 20
#define MINIMUM_STEPPER_PRE_DIR_DELAY 20``` 

To be sure it wasnt that.

Attached configs for reference

[Config.zip](https://github.com/MarlinFirmware/Marlin/files/4616459/Config.zip)

sjasonsmith commented:
Thanks @GhostlyCrowd. It is possible the interrupt priority code isn't working quite as expected. I'll have to fixture up my board to verify they are behaving as expected on the logic analyzer.

mojocorp commented:
@sjasonsmith I've added the M300 support to stm32duino in 1.8.0 that's why I specified the corresponding ststm32 framework as minimum version in Marlin pio config.
Like I said the regression with the M300 was NOT due to the timers but a bug introduced in the malyan lcd. I fixed it since and it's already merged.
I have no issue at all with my M300. Not sure what problem @thinkyhead is refering.

thinkyhead commented:
> probably should have marked this as a draft

The link on the right to do that is small and hard to see, but should be there for the author.

thinkyhead commented:
> I have no issue at all with my M300.

Ah, good to hear. I had been having a hard time figuring out M200v2 and M300 build options, which have been a bit of a moving target lately.

> on the side of the Y axis you can feel it, it's just an ever so slight sputtering
> the speaker will sometimes function correctly

I reverted changes aimed at the GTR, some of which may have also been okay for SKR Pro — the assignment of the PWM timers and the MISO timer. Please review this diff to see if there's anything you would want to try:

https://github.com/MarlinFirmware/Marlin/commit/a06a0c5b88edb923db5b45414226ad20fdcfda1d#diff-443a5c5717b9a729419a33c4a3ded66a

GhostlyCrowd commented:
> > I have no issue at all with my M300.
> 
> Ah, good to hear. I had been having a hard time figuring out M200v2 and M300 build options, which have been a bit of a moving target lately.
> 
> > on the side of the Y axis you can feel it, it's just an ever so slight sputtering
> > the speaker will sometimes function correctly
> 
> @GhostlyCrowd — I reverted changes aimed at the GTR, some of which may have also been okay for SKR Pro — the assignment of the PWM pins/timers and the [MISO pin](https://github.com/MarlinFirmware/Marlin/commit/a06a0c5b88edb923db5b45414226ad20fdcfda1d#diff-443a5c5717b9a729419a33c4a3ded66aL298-R298). Please review this diff to see if there's anything you would want to try:
> 
> [a06a0c5#diff-443a5c5717b9a729419a33c4a3ded66a](https://github.com/MarlinFirmware/Marlin/commit/a06a0c5b88edb923db5b45414226ad20fdcfda1d#diff-443a5c5717b9a729419a33c4a3ded66a)

Once this test is done ill do a fresh clone and test with https://github.com/MarlinFirmware/Marlin/pull/17937

If the micro stutter is not present then that rules it out and ill do more testing with this PR.

sjasonsmith commented:
@mojocorp that is interesting that M300 timers were working for you. They don’t work in the STM32F4 chips without the setMode call I added. Something changed in the 1.8 framework or its underlying HAL that made that necessary.

sjasonsmith commented:
The comment and “Close” butters are very close together on a phone 😕

thisiskeithb commented:
I'm running a test print with this PR on a BTT002/Prusa MK3S (`STM32F4`), so I'll let you know how it goes.

GhostlyCrowd commented:
> I'm running a test print with this PR on a BTT002/Prusa MK3S (`STM32F4`), so I'll let you know how it goes.
> 
> Edit: All good here.

Can i see your configs and compare them to mine. i ant to see why mine is micro stuttering, it appears its a me problem if you're OK, or my specific stepper drivers.

thisiskeithb commented:
> Can i see your configs and compare them to mine. i want to see why mine is micro stuttering, it appears its a me problem if you're OK, or my specific stepper drivers.

I keep them online: [link](https://github.com/thisiskeithb/Marlin/tree/myconfig/MK3S-BTT002-TMC2209). The readme has all my hardware/mods.

GhostlyCrowd commented:
> > Can i see your configs and compare them to mine. i want to see why mine is micro stuttering, it appears its a me problem if you're OK, or my specific stepper drivers.
> 
> I keep them online: [link](https://github.com/thisiskeithb/Marlin/tree/myconfig/MK3S-BTT002-TMC2209). The readme has all my hardware/mods.

Thanks, not difference i can that would cause my micro stuttering. 

Let's assume that's a me problem then.   however, the Speaker not always working is definitely repeatable it happens 2 out of 3 times for me @sjasonsmith. 

I'm going to do some more testing ill let you know. I'm going to make sure the micro stuttering  is not mechanical, as this is what it suggests since @thisiskeithb has no issues liek this.

GhostlyCrowd commented:
So I did a new flash today, I don't know what or why, but the micro stuttering has gone away chalk it up to something odd on my end. 

Speaker is still wonky though. 100% does funky things each time speaker is used it is a dice roll on if it does what its supposed to.

thisiskeithb commented:
> Speaker is still wonky though. 100% does funky things each time speaker is used it is a dice roll on if it does what its supposed to.

Can you clarify this?

Does it not play tones, play them incorrectly, cause other hardware to act up, etc.?

GhostlyCrowd commented:
> > Speaker is still wonky though. 100% does funky things each time speaker is used it is a dice roll on if it does what its supposed to.
> 
> Can you clarify this?
> 
> Does it not play tones, play them incorrectly, cause other hardware to act up, etc.?

At the end of my print I have a melody using M300 that plays when it finishes. Some times it plays fine, other times it just randomly chirps a handful random tones, like its attempting to do the melody but failing obviously, other times chirps once instead of playing the full melody.

thinkyhead commented:
> The comment and “Close” butters are very close together on a phone 😕

Sure, sure, Mister Buttonfingers.

thinkyhead commented:
> It's definitely this PR that does it too I've rolled back and forth a few times did some cubes to test.

Speaker tones are handled by a timer. We rolled our own for DUE but most of them are provided by the framework and use their own timer. If we can determine which timer they use for the buzzer tones, then we can see if it might be conflicting with another or if it has too low a priority…

thinkyhead commented:
Comparing the new framework to the old, I don't see any appreciable difference in the tone handling, except that the new code. Both use the hardware timer associated with the pin. So, if some other function has been reassigned to a related timer, that could also have an impact.

GhostlyCrowd commented:
I've been playing with the speaker.  SO i send a command to it suck as "M300 S2959 P133" and nothing happens then about 60 seconds later it chirps.  Its like its taking 60 seconds to preform ever M300 command.

I'm just sending a simple gcode I made of the Doom theme to test

Without this pr it plays no problem.  I can flash yesterdays Marlin pull and doom.gcode plays fine, flash this PR same configs and i get the above mentions issues.
[M300 Musical Gcode Test.zip](https://github.com/MarlinFirmware/Marlin/files/4624632/M300.Musical.Gcode.Test.zip)

thisiskeithb commented:
I didn't have `SPEAKER` enabled in the BTT002 config (`STM32F4`) with a standard 2004 character LCD controller (`REPRAP_DISCOUNT_SMART_CONTROLLER`), so I reflashed with it enabled and ran a couple `M300` test prints with this PR.

There was no change in tone quality between this PR and standard `bugfix-2.0.x` as both sounded "crunchy": [video](https://youtu.be/ObCygJkAkho).

For comparison against a known-good config, I ran the same test prints on an SKR 1.4 build (`LPC1768`) with a BTT TFT35 E3 (`REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER`) and they sounded a lot cleaner: [video](https://youtu.be/5GauoG5ZF4Y).

I'll have to do some more work to swap LCD controllers to see it's a HAL or controller/speaker quality issue.

[M300 test print.zip](https://github.com/MarlinFirmware/Marlin/files/4624657/M300.test.print.zip)

GhostlyCrowd commented:
> Just to add some more data points:
> 
> I didn't have `SPEAKER` enabled in the BTT002 config (`STM32F4`) with a 2004 character LCD controller (`REPRAP_DISCOUNT_SMART_CONTROLLER`), so I reflashed with it enabled and ran a couple `M300` test prints with this PR.
> 
> There was no change in tone quality between this PR and standard `bugfix-2.0.x` as both sounded "crunchy": [video](https://youtu.be/ObCygJkAkho).
> 
> For comparison against a known-good config, I ran the same test prints on an SKR 1.4 build (`LPC1768`) with a BTT TFT35 E3 V3 (`REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER`) and they sounded a lot cleaner: [video](https://youtu.be/5GauoG5ZF4Y).
> 
> I'll have to do some more work to swap LCD controllers to see it's a HAL or controller/speaker quality issue.
> 
> [M300 test print.zip](https://github.com/MarlinFirmware/Marlin/files/4624657/M300.test.print.zip)

thanks for testing as well, like i said some times it would play fine other times it does nothing and the latest and new symptom is it takes 60 seconds to react to a line of M300 code. 

Also i'm totally stealing your mario tune for my successful print end code.

sjasonsmith commented:
I can probably try to reproduce this on my bench in a few hours and see if I can tell what is happening.

thisiskeithb commented:
I swapped that 2004 character controller over to the `LPC1768` build and it sounds better: [video](https://youtu.be/REhD_bwo2Co).

Something is up with `SPEAKER` on the BTT002/`STM32F4`, but it doesn't produce proper tones with `bugfix-2.0.x` either.

sjasonsmith committed:
Switch TIMER_TONE and TIMER_SERIAL for SKR Pro and GTR
It seems that the framework's Tone class is only working properly with basic timers.

sjasonsmith commented:
There seem to be two different problems, which is why the SKR Pro and BTT002 are behaving differently.

First, it appears that the Tone class in framework 1.8 only wants to work with Basic timers, which limits it to TIM6 and TIM7. Switching the Servo and Tone timers seems to resolve the issue @GhostlyCrowd reported of the speaker basically not working at all.
**This issue also seems to be fixed in the pre-release framework upstream. This appears to be a regression just impacting the 1.8 release.**

Second, the crackly output is you "hearing" the temperature ISR. Tone and temperature ISRs are both at priority 14. If I bump the tone ISR to 13 (by way of the setting a new default timer priority) then the noise mostly goes away. There is a still a tiny bit of noise at high frequencies, but that might be inevitable.

I'll post some updated code soon.

sjasonsmith committed:
Improve tone quality for HAL/STM32 environments
Increase the default timer priority above the temperature ISR, so that steader tones will be produced by the Tone class.

sjasonsmith commented on platformio.ini:
```diff
@@ -38,6 +38,7 @@ lib_deps =
 [common_stm32]
 arduino_ver = >=4.10800
 ststm32_ver = >=6.1
+build_flags = -DTIM_IRQ_PRIO=13
```
I anticipate moving more into this, for other commonalities across all STM32 environments. Some of the USB defines are good candidates.

I didn't do that now to limit the churn in this PR, and because some of the USB defines are going to change in the next framework release anyway. I'm told that release is anticipated within the next couple weeks. Hopefully it will be available through PlatformIO quickly so we will be able to rework the USB and SoftwareSerial configuration then.

GhostlyCrowd replied:
So these changes worked, as you already knew. Just thought I would confirm it on my end for you. 

Works as expected and sounds MUCH better

thisiskeithb commented:
I should have a SKR Mini 1.2 and MKS Robin E3 (both `STM32F1`) test rig by this weekend if you’re looking for more feedback.

sjasonsmith commented:
> I should have an SKR Mini E3 1.2 and MKS Robin E3 (both `STM32F1`) test rig built by this weekend if you’re looking for more feedback.

Those use a different HAL, so these changes don't impact those environments.

thisiskeithb commented:
I gave @sjasonsmith a heads up on Discord, but as soon as I enable `SPEAKER` on the BTT002 after the latest changes in this PR, neither tones, nor steppers work.

Disabling `SPEAKER` or reverting back to `bugfix-2.0.x` fixes the issue.

GhostlyCrowd commented:
Everything is working beautifully for me on the SKR PRo with 2209's on the latest changes to the PR.  It might be in my head but my stepper movement seems quieter and smoother as well.

sjasonsmith commented:
@thisiskeithb you posted earlier that you were printing with these changes. Do your motors work if you go back two commits in this changelist, just before that tone-related stuff?

thisiskeithb commented:
> @thisiskeithb you posted earlier that you were printing with these changes. Do your motors work if you go back two commits in this changelist, just before that tone-related stuff?

Some weird stuff is going on, so maybe I screwed up my copy in git while trying out this PR.

Staring with my `bugfix-2.0.x` copy with SPEAKER `disabled`, everything works:

<details><summary>Version/Log Output</summary>
<pre>Processing BIGTREE_BTT002_BETA (platform: ststm32@5.6.0; board: BigTree_Btt002_beta; framework: arduino)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/BigTree_Btt002_beta.html
PLATFORM: ST STM32 5.6.0 > STM32F407VE (192k RAM. 512k Flash)
HARDWARE: STM32F407VET6 168MHz, 128KB RAM, 512KB Flash
DEBUG: Current (jlink) External (jlink, stlink)
PACKAGES: 
 - framework-arduinoststm32 4.10700.200103 (1.7.0) 
 - toolchain-gccarmnoneeabi 1.70201.0 (7.2.1)
Converting Marlin.ino
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 13 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <LiquidCrystal> 1.5.0
|   |-- <Wire> 1.0
|-- <TMCStepper> 0.6.2
|   |-- <SPI> 1.0
|-- <U8glib-HAL> 0.4.1
|   |-- <Wire> 1.0
|   |-- <SPI> 1.0
|-- <Adafruit MAX31865 library> 1.0.3
|   |-- <SPI> 1.0
|-- <LiquidTWI2>
|   |-- <Wire> 1.0
|-- <Arduino-L6470> 0.8.0
|-- <SPI> 1.0
|-- <Servo> 1.1.2
|-- <IWatchdog> 1.0.0
|-- <Wire> 1.0
Building in release mode</pre>
</details>

Using the same `bugfix-2.0.x` commit, I enabled `SPEAKER` and neither tones, nor steppers work:
<details><summary>Version/Log Output</summary>
<pre>Processing BIGTREE_BTT002_BETA (platform: ststm32@5.6.0; board: BigTree_Btt002_beta; framework: arduino)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/ststm32/BigTree_Btt002_beta.html
PLATFORM: ST STM32 5.6.0 > STM32F407VE (192k RAM. 512k Flash)
HARDWARE: STM32F407VET6 168MHz, 128KB RAM, 512KB Flash
DEBUG: Current (jlink) External (jlink, stlink)
PACKAGES: 
 - framework-arduinoststm32 4.10700.200103 (1.7.0) 
 - toolchain-gccarmnoneeabi 1.70201.0 (7.2.1)
Converting Marlin.ino
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 13 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <LiquidCrystal> 1.5.0
|   |-- <Wire> 1.0
|-- <TMCStepper> 0.6.2
|   |-- <SPI> 1.0
|-- <U8glib-HAL> 0.4.1
|   |-- <Wire> 1.0
|   |-- <SPI> 1.0
|-- <Adafruit MAX31865 library> 1.0.3
|   |-- <SPI> 1.0
|-- <LiquidTWI2>
|   |-- <Wire> 1.0
|-- <Arduino-L6470> 0.8.0
|-- <SPI> 1.0
|-- <Servo> 1.1.2
|-- <IWatchdog> 1.0.0
|-- <Wire> 1.0
Building in release mode</pre>
</details>

It seems that enabling `SPEAKER` in any commit causes tones & steppers to not work. I'll do some more testing.

thisiskeithb commented:
I re-ran tests with my [BTT002 config](https://github.com/thisiskeithb/Marlin/tree/myconfig/MK3S-BTT002-TMC2209) based on the latest `bugfix-2.0.x` (64876f7) and here's what I found:
- My standard config + `SPEAKER` disabled = Motors work, tones don't (expected)
- My standard config + `SPEAKER` enabled = Motors & tones don't work (unexpected)
- This PR + `SPEAKER`disabled = Motors work, tones don't (expected)
- This PR + `SPEAKER` enabled = Motors & tones don't work (unexpected)
- This PR reverted to cec5f3c + `SPEAKER` disabled = Motors work, tones don't (expected)
- This PR reverted to cec5f3c + `SPEAKER` enabled = Motors & tones don't work (unexpected)

I don't know how I got tones working before. Also, `SPEAKER` is disabled in the standard BTT002 config, so we can leave it disabled.

sjasonsmith commented:
@thisiskeithb Can you change the variant.h file for the BTT002 so that it uses the same timers as the SKR Pro? I didn't notice any conflicts there, so hopefully that will be ok. Hopefully that will fix whatever is going wrong.

thisiskeithb commented:
> @thisiskeithb Can you change the variant.h file for the BTT002 so that it uses the same timers as the SKR Pro?

Motors & tones work with `SPEAKER` enabled after that change to the BTT002 variant file.

Having said that, `TIM2` is already assigned to `TIMER_SERVO`:
https://github.com/MarlinFirmware/Marlin/blob/64876f77817bc768f5e26b0d2ff1268f35c88b91/buildroot/share/PlatformIO/variants/BIGTREE_BTT002/variant.h#L251-L252

There's no dedicated servo pins on this board, so maybe that can be ignored or changed to something else?

sjasonsmith commented:
@thisiskeithb It is probably ok to change all three timers in the variant.h to match the SKR Pro. That servo timer will be used if anyone adds a servo or BLTouch to any unused pin. I’ll push a change tonight, or if you’ve tested changes feel free to send a pull request to my branch.

sjasonsmith commented:
Oh yeah, tone and step were both using timer 6. That should have been an obvious thing to check. I’ll post an update with a check added to do a basic for conflicting timers at build time. It won’t detect conflicts with PWM pins, but could have caught this.

thisiskeithb commented:
> It is probably ok to change all three timers in the variant.h to match the SKR Pro.

I've updated the serial, servo, and tone timers in the BTT002 variant file to match the SKR Pro and ran a test print. Everything works, including much improved tone quality.

I can't easily add a BLTouch to test if servos work right now, but this board was designed for a PINDA probe/stock MK3S.

GhostlyCrowd commented:
Ive got about 10 hours worth of printing on this pr with its latest changes, working like a dream on the SKR Pro, Bltouch,  2209's, speaker works and no longer sounds like a baby yak.

Bob-the-Kuhn commented:
Thanks for getting this working.  This will allow me to move forward on my USB composite work.

I have a buggy version working with the current library.  Switching to the latest library will allow me to produce and test code that can be merged into the STM32 library.

sjasonsmith commented:
@viper93458, does it work properly with the latest state of bugfix-2.0.x?

If so, can you both describe your EEPROM hardware and provide configs, pin files, etc.?

viper93458 commented:
Yes it does as of 5/12/2020 bugfix.

No changes outside of config and config_adv.

I use this in my config to setup the EEPROM:

//Added external EEPROM
#define I2C_EEPROM // Use external EEPROM Module (e.g. AT24C256) connected to I2C Pins as outlined at https://www.instructables.com/id/BigTreeTech-SKR-Pro-V11-Adding-a-EEPROM/
#ifdef I2C_EEPROM
  #define USE_SHARED_EEPROM 1 // Use Platform-independent Arduino functions for I2C EEPROM
  #undef E2END                // Defined in Arduino Core STM32 to be used with EEPROM emulation. However a real EEPROM to be used
  #define E2END 0x7FFF        // EEPROM end address AT24C256 (32kB)
#endif
//Added external EEPROM

I am using the Robotdyn module and the wiring and address setting from this intractable:
https://www.instructables.com/id/BigTreeTech-SKR-Pro-V11-Adding-a-EEPROM/?fbclid=IwAR2uKRGzvIFZJv8fiw6MxmGGHm8PG-gGMObgHNx5tX_Bklcl2R1uMyO01qQ

-William

sjasonsmith commented:
Thanks @viper93458. I've never used them, but I have some similar modules that arrived a few days ago. I'll try it out tonight.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/17970

thinkyhead commented:
The code for `I2C_EEPROM` + `USE_SHARED_EEPROM` wants to use the generic i2c interface (in `HAL/shared`). If i2c EEPROM is broken by this PR, it's possibly some i2c-related change in the framework, but maybe some other change is conflicting with i2c.

randellhodges commented:
I have that very module and had it working prior to this PR, so I'll give it a shot once I get this PR merged and report back.

randellhodges commented:
I can confirm.

This is was my configuration rebased on top of bugfix-2.0.x from a few minutes ago:
https://github.com/randellhodges/Marlin/commit/166151b06d63a34eb599501911baae373e3c1842
And that worked.

I then adding @sjasonsmith fork as a remote, fetched it and then rebase on top of that. Built and flashed.

`M503` works and I verified it has the changes I made and saved prior to merging this PR. But as soon as I do an `M500` (I changed `M900 K1.5` then executed `M500`)

```
SENDING:M500
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Settings Stored (647 bytes; crc 56674)
Error:EEPROM datasize error.
[ERROR] Error:EEPROM datasize error.
```

sjasonsmith committed:
Fix BTT002 timer conflict with SPEAKER enabled
This now uses the same timers as the SKR Pro, which does not appear to conflict with other board resources.

sjasonsmith commented:
> I’ll post an update with a check added to look for conflicting timers at build time.

Adding the check wasn't as simple as I thought it would be, since the timers defined in variant.h aren't simple numbers. They dereference into pointers which `static_assert` doesn't seem to be happy with. It's probably solvable, but making sense of the I2C issue seems more important.

randellhodges commented:
This is a problem with Arduino_Core_STM32 1.8.0 Wire.cpp/Wire.h
https://github.com/stm32duino/Arduino_Core_STM32/tree/1.8.0/libraries/Wire/src

I replaced Wire.cpp and Wire.h with version 1.9.0 and my eeprom works again.

\.platformio\packages\framework-arduinoststm32\libraries\Wire\src\Wire.cpp
\.platformio\packages\framework-arduinoststm32\libraries\Wire\src\Wire.h

Replace those with the 2 from here:
https://github.com/stm32duino/Arduino_Core_STM32/tree/1.9.0/libraries/Wire/src

sjasonsmith commented:
@randellhodges ok, so they released 1.9! That's great, now we just have to wait until it is available through PlatformIO. I have no idea what gates that, but it took about two months for version 1.8 to show up :(

randellhodges commented:
My eyes are still bleeding chasing that rabbit. I didn't read your first post, so it took forever before I found the correct GitHub, and the only thing that clued me in on 1.8 was your branch name.

I might have missed the obvious, but I don't see the relationship between version 1.8 and the versions you specify in platformio.

I'm a newb, so maybe I just overlooked it. If I didn't, they sure don't make this easy.

thinkyhead commented:
I see that the new i2c Wire code makes a bunch of data members that used to be `static` class members into regular instance members. So instead of i2c sharing these values, they are now set on a per-instance basis. The variables that used to be static are these:
```cpp
// Initialize Class Variables //////////////////////////////////////////////////
uint8_t *TwoWire::rxBuffer = nullptr; // Init by begin() in 1.9.x
uint8_t TwoWire::rxBufferAllocated = 0; // Init by begin() in 1.9.x
uint8_t TwoWire::rxBufferIndex = 0; // Set by onReceiveService in 1.9.x
uint8_t TwoWire::rxBufferLength = 0; // Set by onReceiveService in 1.9.x

uint8_t TwoWire::txAddress = 0; // Init by begin() in 1.9.x
uint8_t *TwoWire::txBuffer = nullptr; // Init by begin() in 1.9.x
uint8_t TwoWire::txBufferAllocated = 0; // Init by begin() in 1.9.x
uint8_t TwoWire::txBufferIndex = 0; // Set by onRequestService in 1.9.x
uint8_t TwoWire::txBufferLength = 0; // Set by onRequestService in 1.9.x

uint8_t TwoWire::transmitting = 0;
void (*TwoWire::user_onRequest)(void); // Init by begin() in 1.9.x
void (*TwoWire::user_onReceive)(int);
```

Also the `onReceiveService` / `onRequestService` methods take an object instead of a buffer pointer/length and void, and use the in instance `user_onReceive` / `user_onRequest` pointers.

So, the long and short is that every distinct `TwoWire` instance needs to call `begin` and can no longer rely on others having called `begin`.

sjasonsmith commented:
At this point I am thinking we should skip 1.8 and wait for 1.9 to be available through PlatformIO, although I have no idea how long that will be.

randellhodges commented:
https://github.com/stm32duino/Arduino_Core_STM32/releases

STSTM32 5.7.0 (released 7 months ago) brought in 1.7 (released 7 months ago)
STSTM32 6.1.0 (released about 1 month ago) brought in 1.8 (released 5 months ago)
? 1.9 (released 1 day ago...)

I think we'll have a long wait....

A few bullet points caught my attention. I have no idea if they actually mean anything, they have all the keywords there...

A bunch of timer stuff. I wonder if this has any impact on BLTouch...
`HardwareTimer: start timer when only update interrupt needed (#841)`
`analogWrite: Configure HardwareTimer mode only if not previously done (avoid glitch)`

Flash EEPROM
`Fix EEPROM hardfault: eeprom_buffer should be 8 bytes aligned (#931)`
Does this mean we should change the wear leveling buffer we are using to be 8 bytes aligned?

Marlin's:
https://github.com/MarlinFirmware/Marlin/blob/847bdeecafd14ea611af0bda723ebe99c8d2d29b/Marlin/src/HAL/STM32/eeprom_flash.cpp#L90

Arduino_Core_STM32:
https://github.com/stm32duino/Arduino_Core_STM32/blob/17fd01f9ff569cc55de975da6a5742a8aab6d004/libraries/SrcWrapper/src/stm32/stm32_eeprom.c#L145

sjasonsmith commented:
> A bunch of timer stuff. I wonder if this has any impact on BLTouch...
HardwareTimer: start timer when only update interrupt needed (#841)
analogWrite: Configure HardwareTimer mode only if not previously done (avoid glitch)

This is an example of a regression in 1.8 that was fixed in 1.9. I had to add calls to `HardwareTimer::setMode` in this PR to make it work in 1.8, but that was not necessary in 1.7 or 1.9.

I actually don't think there is a lot to gain from going to 1.8, especially if it breaks I2C, which then unbreaks in 1.9. There are multiple items in 1.9 that make things cleaner in Marlin, we just don't know when we will be able to consume it.

@randellhodges, did you write multiple times in 1.9 to make sure it was reliable? I found that in 1.8 it actually does work sometimes, so is is possible you just got lucky with 1.9.

randellhodges commented:
I thought I did a few times. I admit I didn't extensively test it, but with
1.8 I never got a single save.

With the wire files from 1.9 I think I tried it a few times and proclaimed
success.

On Sat, May 16, 2020, 7:27 PM Jason Smith <notifications@github.com> wrote:

> A bunch of timer stuff. I wonder if this has any impact on BLTouch...
> HardwareTimer: start timer when only update interrupt needed (#841
> <https://github.com/MarlinFirmware/Marlin/pull/841>)
> analogWrite: Configure HardwareTimer mode only if not previously done
> (avoid glitch)
>
> This is an example of a regression in 1.8 that was fixed in 1.9. I had to
> add calls to HardwareTimer::setMode in this PR to make it work in 1.8,
> but that was not necessary in 1.7 or 1.9.
>
> I actually don't think there is a lot to gain from going to 1.8,
> especially if it breaks I2C, which then unbreaks in 1.9. There are multiple
> items in 1.9 that make things cleaner in Marlin, we just don't know when we
> will be able to consume it.
>
> @randellhodges <https://github.com/randellhodges>, did you write multiple
> times in 1.9 to make sure it was reliable? I found that in 1.8 it actually
> does work sometimes, so is is possible you just got lucky with 1.9.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/17970#issuecomment-629724256>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB26S22XDNVLHDIO2M64HJTRR4VOXANCNFSM4M6PVFHA>
> .
>

thinkyhead commented:
If there are no bodges possible for this, then it may as well be closed. However, since the changes here are going to apply to 1.9, I will leave it open for the next several weeks. Of course, keep us posted if a 1.8.1 patch fixes things sooner!

swilkens commented:
> > I should have an SKR Mini E3 1.2 and MKS Robin E3 (both `STM32F1`) test rig built by this weekend if you’re looking for more feedback.
> 
> Those use a different HAL, so these changes don't impact those environments.

Perhaps a silly question, but can this be consolidated to the same HAL? Looking at both frameworks, they seem to support F1 as well as F4?

thinkyhead commented:
> Perhaps a silly question, but can this be consolidated to the same HAL?

If the framework that the HAL is built on can support everything that is needed for the board, then it is possible to switch to one of the other HALs. Just open up `platformio.ini` and make a copy of the existing environment, tweak it to use the HAL you want to try, and then build under the new environment to see what happens. AFAIK the variants files aren't heavily tailored to a particular Marlin HAL, but each Marlin HAL is tailored to a certain framework.

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into PR/STM32_1.8

thinkyhead committed:
Avoid duplication of build flags

thinkyhead commented:
I received an email about STM32 support in PlatformIO going forward. One of the big changes is the dropping of support for Maple core, so all our PIO targets need to be built on STM32Duino.

> Hi Scott, my name is Valerii and I'm responsible for the official development platforms at PlatformIO.
>
> I'm writing to let you know that we plan to release a new major version of the ststm32 platform that will include some breaking changes that might affect the Marlin project:
> 
> We're going to drop support for the Maple core. It'll be superseded by the official STM32Duino for all boards (including genericSTM32F1* that's used in Marlin). I'm not sure what are possible consequences of this change for your project, but it looks like there are differences between these cores in terms of API.
>
> tool-stm32duino is going to be updated as well (new package structure, updated binaries, etc). A new tool "STM32CubeProgrammer" will be used for "serial" and "DFU" upload protocols and unfortunately, there is no binary for ARM architecture so it looks like these protocols won't be functional on small computers like Raspberry Pi, etc
>
> Minor change: a new board fysetc_s6 that might conflict with an identical board in your "platformio.ini".
>
> To handle this new release, you can:
> Create a separate branch with the development version of ststm32 platform to evaluate new changes, just to make sure that Marlin can be built with the upcoming release.
> Fix the platform version to 6.1.0 in the platformio.ini file (maybe just temporarily, to make sure Marlin stays compilable)

sjasonsmith commented:
> I received an email about STM32 support in PlatformIO going forward. One of the big changes is the dropping of support for Maple core, so all our PIO targets need to be built on STM32Duino.

OK, I guess that lessens the urgency of some TONE-related changes I was working on for the STM32F1 HAL, if everything's just going to have to move to the other HAL anyway.

I'll see if I can figure out how to pull down their development version and see how it goes.

> Minor change: a new board fysetc_s6 that might conflict with an identical board in your "platformio.ini".

This concerns me. It seems like we end up having to make changes in variants all the time for Marlin, so having the variants already in the framework feels more like a hindrance than a help. I don't know whether it is really going to be practical to use the variants from the framework, or if we will just have to maintain our own Marlin variants anyway, such as `MARLIN_FYSETC_S6`.

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into PR/STM32_1.8

viper93458 commented:
I am presently using the branch that this PR is derived from. I tried to use the current bugfix 2.0 but its giving my bltouch some troubles. I started using this branch to get around boot loops i was having with my skr pro previously. 

By chance is the branch that this PR is coming from fully up to date so i can use it and have all recent bugfix merges too? I apologize for not knowing how to check that myself. I know further work here is pending stm32 1.9 framework still (and probably lots more stuff).

-William

sjasonsmith commented:
@viper93458 the boot loops were caused by too new of an arduino framework being used. New versions of bugfix should be preventing that, but I haven't actually tried things in a couple weeks.

This PR is currently stalled. There appeared to be some issues related to the I2C EEPROM support that prevented it from moving forward.

At this point we might need some guidance from @thinkyhead regarding how he would like to approach this.

viper93458 commented:
EEPROM is working great on my SKR pro With your branch as of 5/9/2020. I know there were conversations above about it. I have bltouch issues with current bug fix which don’t happen with your branch so I figured the various changes must be affecting it and wa hoping this branch along with current merges would work together to bring me more current. :)

sjasonsmith commented:
@viper93458 are you using an I2C EEPROM? Default will be either flash or SD EEPROM.

I was able to reproduce the issues others reported with an I2C EEPROM, although I did not spend any time debugging it.

This branch should really not have any advantages over the current bugfix-2.0.x for an SKR Pro. If there are new BLTouch issues in the bugfix-2.0.x branch, that needs investigated. Most likely something got messed up with timer assignments or priorities.

viper93458 commented:
Yes, i2c eeprom as reported much earlier in this thread.

viper93458 commented:
> 
> 
> @viper93458 are you using an I2C EEPROM? Default will be either flash or SD EEPROM.
> 
> I was able to reproduce the issues others reported with an I2C EEPROM, although I did not spend any time debugging it.
> 
> This branch should really not have any advantages over the current bugfix-2.0.x for an SKR Pro. If there are new BLTouch issues in the bugfix-2.0.x branch, that needs investigated. Most likely something got messed up with timer assignments or priorities.

Agreed, I submitted a bug report: https://github.com/MarlinFirmware/Marlin/issues/18372

-William

thinkyhead committed:
Merge branch 'bugfix-2.0.x' into PR/STM32_1.8

thinkyhead committed:
Update platformio.ini

thinkyhead committed:
Update platformio.ini

thinkyhead committed:
Update platformio.ini

sjasonsmith commented:
@thinkyhead 
A new platform package has been released for 4.109. My earlier experiments showed this should eliminate several of the workarounds 4.108 required, and may resolve the outstanding I2C EEPROM issue.

I will try updating to the newer framework and update this, for inclusion after the 2.0.6 release. I'm not sure yet whether I will be able to update just the framework version without also updating to a newer platform which might break things.

thinkyhead commented:
> …without also updating to a newer platform…

Good luck! I look forward to your results.

sjasonsmith commented:
I had a FYSETC S6 booting and at least LCD, fans, and Temp ISRs working using ststm32 7.0 with the new framework. Unfortunately I seem to have burned the board up when I added a TMC2208 to test SoftwareSerial. Oops.

I'll have to switch back to my SKR Pro and hope for better results there. I should get to it in the next few days.

sjasonsmith commented:
Rather than continuing this PR, I am choosing to drop this one and create a new PR (18496) for the upgrade to the newer framework.

The changes needed for versions 1.8 and 1.9 were mostly unique, so it didn't seem to make sense to build it on top of this.

thinkyhead commented:
Duplcate of #18496

thinkyhead commented:
Duplicate of #18496
